tracee-rules/benchmark/benchmark_test.go

```diff
@@ -0,0 +1,143 @@
+		err = codeInjectSig.OnEvent(triggerCodeInjectorPtraceEvent)
+		require.NoError(b, err)
+	}
+}
+
+func BenchmarkOnEventCodeInjectionGoRule(b *testing.B) {
+	codeInjectSig := golang.NewCodeInjectionSignature()
+	err := codeInjectSig.Init(ignoreFinding)
+	require.NoError(b, err)
+
+	for i := 0; i < b.N; i++ {
+		err := codeInjectSig.OnEvent(triggerCodeInjectorPtraceEvent)
+		require.NoError(b, err)
+	}
+}
+
+func BenchmarkEngineWithCodeInjectionRegoRule(b *testing.B) {
+	// Prepare signatures
+	codeInjectSig, err := rego.NewCodeInjectionSignature()
+	require.NoError(b, err)
+
+	sigs := []types.Signature{
+		codeInjectSig,
+	}
+
+	for i := 0; i < b.N; i++ {
+		// Produce events
+		eventsCh := make(chan types.Event, inputEventsCount)
+		ProduceEventsInMemory(eventsCh, inputEventsCount)
+
+		inputs := engine.EventSources{
+			Tracee: eventsCh,
+		}
+		output := make(chan types.Finding, inputEventsCount)
+
+		e := engine.NewEngine(sigs, inputs, output, os.Stderr)
+		// Start rules engine and wait until all events are processed
+		e.Start(waitForEventsProcessed(eventsCh))
+
+		b.Logf("Test is done with %d findings", len(output))
+	}
+}
+
+func BenchmarkEngineWithCodeInjectionGoRule(b *testing.B) {
+	// Prepare signatures
+	codeInjectSig := golang.NewCodeInjectionSignature()
+
+	sigs := []types.Signature{
+		codeInjectSig,
+	}
+
+	for i := 0; i < b.N; i++ {
+		// Produce events
+		eventsCh := make(chan types.Event, inputEventsCount)
+		ProduceEventsInMemory(eventsCh, inputEventsCount)
+
+		inputs := engine.EventSources{
+			Tracee: eventsCh,
+		}
+		output := make(chan types.Finding, inputEventsCount)
+
+		e := engine.NewEngine(sigs, inputs, output, os.Stderr)
+		// Start rules engine and wait until all events are processed
+		e.Start(waitForEventsProcessed(eventsCh))
+
+		b.Logf("Test is done with %d findings", len(output))
+	}
+}
+
+func BenchmarkEngineWithMultipleRegoAndGoRules(b *testing.B) {
+	// Prepare signatures
+	codeInjectionRegoSig, _ := rego.NewCodeInjectionSignature()
+	antiDebuggingRegoSig, _ := rego.NewAntiDebuggingSignature()
+
+	sigs := []types.Signature{
+		codeInjectionRegoSig,
+		antiDebuggingRegoSig,
+
+		golang.NewCodeInjectionSignature(),
+		golang.NewAntiDebuggingSignature(),
+	}
+
+	for i := 0; i < b.N; i++ {
+		// Produce events
+		eventsCh := make(chan types.Event, inputEventsCount)
+		ProduceEventsInMemory(eventsCh, inputEventsCount)
+
+		inputs := engine.EventSources{
+			Tracee: eventsCh,
+		}
+		output := make(chan types.Finding, inputEventsCount*len(sigs))
+
+		e := engine.NewEngine(sigs, inputs, output, os.Stderr)
+		// Start rules engine and wait until all events are processed
+		e.Start(waitForEventsProcessed(eventsCh))
+
+		b.Logf("Test is done with %d findings", len(output))
+	}
+}
+
+func waitForEventsProcessed(eventsCh chan types.Event) chan bool {
+	done := make(chan bool, 1)
+	go func() {
+		for {
+			if len(eventsCh) == 0 {
+				done <- true
+				return
+			}
+			time.Sleep(100 * time.Millisecond)
```

Comment:
Reducing this wait down exposes a bug in the code: https://github.com/aquasecurity/tracee/issues/777

```suggestion
			time.Sleep(1 * time.Millisecond)
```

If patched the numbers look a lot different:

```

benchmark                                     iter           time/iter     bytes alloc             allocs
---------                                     ----           ---------     -----------             ------
BenchmarkOnEventCodeInjectionRegoRule-8        100      56327.00 ns/op      23318 B/op      476 allocs/op
BenchmarkOnEventCodeInjectionGoRule-8          100        813.80 ns/op        608 B/op        4 allocs/op
BenchmarkEngineWithCodeInjectionRegoRule-8     100   12355021.00 ns/op    5762957 B/op   112738 allocs/op
BenchmarkEngineWithCodeInjectionGoRule-8       100    1449184.00 ns/op     550480 B/op     2550 allocs/op
BenchmarkEngineWithMultipleRegoAndGoRules-8    100   15819138.00 ns/op   11116686 B/op   214733 allocs/op
ok      github.com/aquasecurity/tracee/tracee-rules/benchmark   3.179s
```

Reply:
IOW waiting here makes the numbers artificially skew larger than they should be. 

Somewhat related to this:
The way to get around the bug is to see the above numbers is to not close the channel here https://github.com/aquasecurity/tracee/blob/ef7e96ace8592fe6eb2333008391f0bfb9b6ce8e/tracee-rules/engine/engine.go#L72 – Ideally this channel close should happen from the creator (test program) and not within the engine itself.

Reply:
Actually I take this back. The numbers were different for another reason. If you add `b.ResetTimer()` to all the appropriate places after the setup is done the numbers are much better:

```
PASS
benchmark                                     iter            time/iter     bytes alloc             allocs
---------                                     ----            ---------     -----------             ------
BenchmarkOnEventCodeInjectionRegoRule-8        100       45904.00 ns/op      17971 B/op      353 allocs/op
BenchmarkOnEventCodeInjectionRegoRule-8        100       52121.00 ns/op      17703 B/op      352 allocs/op
BenchmarkOnEventCodeInjectionRegoRule-8        100       40360.00 ns/op      17645 B/op      352 allocs/op
BenchmarkOnEventCodeInjectionRegoRule-8        100       41343.00 ns/op      17628 B/op      352 allocs/op
BenchmarkOnEventCodeInjectionRegoRule-8        100       33396.00 ns/op      17665 B/op      352 allocs/op
BenchmarkOnEventCodeInjectionGoRule-8          100         554.00 ns/op        560 B/op        4 allocs/op
BenchmarkOnEventCodeInjectionGoRule-8          100         567.60 ns/op        560 B/op        4 allocs/op
BenchmarkOnEventCodeInjectionGoRule-8          100         564.50 ns/op        560 B/op        4 allocs/op
BenchmarkOnEventCodeInjectionGoRule-8          100         568.20 ns/op        560 B/op        4 allocs/op
BenchmarkOnEventCodeInjectionGoRule-8          100         560.30 ns/op        560 B/op        4 allocs/op
BenchmarkOnEventCodeInjectionWASMRule-8        100       28182.00 ns/op       1751 B/op       26 allocs/op
BenchmarkOnEventCodeInjectionWASMRule-8        100       21746.00 ns/op       1751 B/op       26 allocs/op
BenchmarkOnEventCodeInjectionWASMRule-8        100       21070.00 ns/op       1751 B/op       26 allocs/op
BenchmarkOnEventCodeInjectionWASMRule-8        100       20489.00 ns/op       1751 B/op       26 allocs/op
BenchmarkOnEventCodeInjectionWASMRule-8        100       20153.00 ns/op       1751 B/op       26 allocs/op
```

WASM is nearly half! 🚀

Reply:
Great find!
just to correct a small comment you made

> Ideally this channel close should happen from the creator (test program) and not within the engine itself

this close is in the right place, engine IS the creator:
https://github.com/aquasecurity/tracee/blob/ef7e96ace8592fe6eb2333008391f0bfb9b6ce8e/tracee-rules/engine/engine.go#L35